### PR TITLE
Clip preprocessors, Habitat usability improvements, and minor fixes.

### PR DESCRIPTION
**Pull request by Lucaweihs**
This PR adds:
1. The clip preprocessor we used for the embodied clip paper (https://arxiv.org/abs/2111.09888) along with relevant RoboTHOR/Habitat experiments.
2. Reworking our point/object nav experiment configs so that we don't use multiple inheritance (as this was causing some surprising behavior).
3. Some bug fixes associated with the new storage system (in particular, saved models didn't store everything needed to ensure the tensorboard picked up where it left off when loading a model).

### CODE REVIEW

File: projects/objectnav_baselines/experiments/robothor/objectnav_robothor_rgb_resnet18gru_ddppo.py [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+from projects.objectnav_baselines.experiments.robothor.objectnav_robothor_base import (
+    ObjectNavRoboThorBaseConfig,
+)
+from projects.objectnav_baselines.mixins import (
+    ResNetPreprocessGRUActorCriticMixin,
+    ObjectNavPPOMixin,
+)
+
+
+class ObjectNavRoboThorRGBPPOExperimentConfig(ObjectNavRoboThorBaseConfig,):
```
Comment (jordis-ai2) [inline projects/objectnav_baselines/experiments/robothor/objectnav_robothor_rgb_resnet18gru_ddppo.py:20]: I guess the comma after the base config class is no longer needed.

Other than that, I'll use this comment to point out that, as much as I like having most of the config in a single file, I'm not sure what the problem with multiple inheritance was. As long as we follow a certain order, the latest overload should be the one in use, right? Just curious about what the issue was :)

File: projects/pointnav_baselines/experiments/habitat/pointnav_habitat_base.py [hunk 2]
```diff
@@ -96,22 +127,55 @@ class PointNavHabitatBaseConfig(PointNavBaseConfig, ABC):
...
-    NUM_TRAIN_PROCESSES = max(5 * torch.cuda.device_count() - 1, 4)
-    NUM_VAL_PROCESSES = 1
-    NUM_TEST_PROCESSES = 10
+    ACTION_SPACE = gym.spaces.Discrete(len(PointNavTask.class_action_names()))
 
-    TRAINING_GPUS = list(range(torch.cuda.device_count()))
-    VALIDATION_GPUS = [torch.cuda.device_count() - 1]
-    TESTING_GPUS = [torch.cuda.device_count() - 1]
+    DEFAULT_NUM_TRAIN_PROCESSES = (
+        max(5 * torch.cuda.device_count(), 4) if torch.cuda.is_available() else 1
```
Comment (jordis-ai2) [inline projects/pointnav_baselines/experiments/habitat/pointnav_habitat_base.py:133]: Is this `max` operation doing anything? (I, perhaps wrongly, think it doesn't)

**Review by jordis-ai2 — APPROVED** [review]
LGTM 👍 I'm not 100% understanding the need for getting rid of the multiple inheritance in experiment configs, but I do like the new design, so feel free to go ahead. Other than that, I just have minor comments. Great job!

File: allenact/embodiedai/models/visual_nav_models.py [hunk 3]
```diff
@@ -70,11 +70,12 @@ def create_state_encoders(
         num_rnn_layers: int,
         rnn_type: str,
         add_prev_actions: bool,
+        add_prev_action_null_token: bool,
         trainable_masked_hidden_state=False,
     ):
         rnn_input_size = obs_embed_size
         self.prev_action_embedder = FeatureEmbedding(
-            input_size=self.action_space.n,
+            input_size=self.action_space.n + add_prev_action_null_token,
```
[the remark this answers is not in the document]
Comment (Lucaweihs) [reply allenact/embodiedai/models/visual_nav_models.py:78]: Done and done, I totally agree that the order should have `add_prev_action_null_token` first :)

File: allenact/embodiedai/preprocessors/resnet.py [hunk 4]
```diff
@@ -53,7 +53,7 @@ def __init__(
         torchvision_resnet_model: Callable[..., models.ResNet] = models.resnet18,
         device: Optional[torch.device] = None,
         device_ids: Optional[List[torch.device]] = None,
-        **kwargs: Any
+        **kwargs: Any,
```
[the remark this answers is not in the document]
Comment (Lucaweihs) [reply allenact/embodiedai/preprocessors/resnet.py:56]: Interesting, I don't know if I 100% follow what the design difference is that you prefer (is it that there is a `clip_plugin`?), let's chat about it.

File: allenact/utils/inference.py [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+        self.actor_critic.to(device=self.device)
+        if self.memory is not None:
+            self.memory.to(device=self.device)
+        self.sensor_preprocessor_graph.to(self.device)
+
+        self.rollout_storage.to(self.device)
+        self.rollout_storage.set_partition(index=0, num_parts=1)
+
+    @classmethod
+    def from_experiment_config(
```
[the remark this answers is not in the document]
Comment (Lucaweihs) [reply allenact/utils/inference.py:42]: Yah, basically the idea here is to make it easier for people to run inference with a model they have trained without explicitly running things through the allenact runner/engine. Here's a little script I threw together to test this out locally:
```python
import torch

from allenact.utils.inference import InferenceAgent
from projects.objectnav_baselines.experiments.robothor.objectnav_robothor_rgbd_resnet18gru_ddppo import (
    ObjectNavRoboThorRGBPPOExperimentConfig,
)

if __name__ == "__main__":
    exp_config = ObjectNavRoboThorRGBPPOExperimentConfig()
    agent = InferenceAgent.from_experiment_config(
        exp_config=ObjectNavRoboThorRGBPPOExperimentConfig(),
        device=torch.device("cpu"),
        checkpoint_path="pretrained_model_ckpts/robothor-objectnav-challenge-2021/Objectnav-RoboTHOR-RGBD-ResNetGRU-DDPPO/2021-02-09_22-35-15/exp_Objectnav-RoboTHOR-RGBD-ResNetGRU-DDPPO_0.2.0a_300M__stage_00__steps_000170207237.pt",
    )

    task_sampler = exp_config.make_sampler_fn(
        **exp_config.valid_task_sampler_args(process_ind=0, total_processes=1)
    )

    for _ in range(10):
        agent.reset()

        task = task_sampler.next_task()
        observations = task.get_observations()

        while not task.is_done():
            action = agent.act(observations=observations)
            observations = task.step(action).observation

        print({k: v for k, v in task.metrics().items() if k != "task_info"})
```
It would probably be a good idea to add a test for this...

File: allenact/utils/experiment_utils.py [hunk 6]
```diff
@@ -1020,12 +1022,20 @@ def load_state_dict(self, state_dict: Dict[str, Any]):
...
                 " that logging for the off-policy component will not behave as it previously did."
+                " Additionally, while the total step count will remain accurate, step counts"
+                " associated with losses will be reset to step 0."
             )
 
         for ps, stage_info in zip(self.pipeline_stages, state_dict["stage_info_list"]):
             ps.early_stopping_criterion_met = stage_info["early_stopping_criterion_met"]
             ps.steps_taken_in_stage = stage_info["steps_taken_in_stage"]
 
+            if "storage_uuid_to_steps_taken_in_stage" in stage_info:
```
[the remark this answers is not in the document]
Comment (Lucaweihs) [reply allenact/utils/experiment_utils.py:1033]: Haha thanks :) basically what's happening now is that each of the different types of storage track how many "steps" they have taken (e.g. the amount of offpolicy storage steps might be different than the number of onpolicy steps). As each storage gets its own tensorboard section, this makes it possible to track the training loss metrics corresponding to that storage using it's definition of "steps." The above saves each storages' step count so that these can be correctly displayed in the tensorboard when restarting a training pipeline.

[the remark this answers is not in the document]
Comment (jordis-ai2) [reply allenact/utils/experiment_utils.py:1033]: 👍

File: allenact_plugins/clip_plugin/clip_preprocessors.py [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+import numpy as np
+import torch
+import torch.nn as nn
+from clip.model import CLIP
+
+from allenact.base_abstractions.preprocessor import Preprocessor
+from allenact.utils.misc_utils import prepare_locals_for_super
+
+
+class ClipResNetEmbedder(nn.Module):
```
[the remark this answers is not in the document]
Comment (Lucaweihs) [reply allenact_plugins/clip_plugin/clip_preprocessors.py:14]: Good point, let's keep it separate for now and see if anyone finds it useful to have as a separate module, if not we can move its functionality into the `ClipResNetPreprocessor`.

File: allenact/utils/inference.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (jordis-ai2) [reply allenact/utils/inference.py:42]: Really cool! I agree it would be nice to have a small test.

File: projects/objectnav_baselines/experiments/robothor/objectnav_robothor_rgb_resnet18gru_ddppo.py (hunk 1 above)
Comment (Lucaweihs) [reply projects/objectnav_baselines/experiments/robothor/objectnav_robothor_rgb_resnet18gru_ddppo.py:20]: Very fair question :) the problem I was having was that the order of method resolution in multiple inheritance was causing me some issues when I tried to reuse some experiment configs across multiple projects. In particular, there was a `Base` config who returned `tuple()` for its set of preprocessors and I wanted to mix this with a `CLIPPreprocessorMixin` config that defined the clip preprocesor. The issue then was the if I put `Base` first, I wouldn't get the clip preprocessors (as it would return `tuple()`) but if I put `CLIPPreprocessorMixin` then it would go through that preprocessor's methods first and find that one of them had a `raise NotImplementedError`.

All this can be fixed but it meant having to modify the experiment configs of other projects and constantly thinking about how things were being resolved which was a bit painful. This new way requires more boilerplate but I like that things are fairly obvious.

File: projects/pointnav_baselines/experiments/habitat/pointnav_habitat_base.py (hunk 2 above)
Comment (Lucaweihs) [reply projects/pointnav_baselines/experiments/habitat/pointnav_habitat_base.py:133]: Haha, yes you're right.

File: allenact/utils/inference.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (Lucaweihs) [reply allenact/utils/inference.py:43]: Added (and it found a bug where experiment configs without sensor preprocessor graphs would break).

File: projects/objectnav_baselines/experiments/robothor/objectnav_robothor_rgb_resnet18gru_ddppo.py (hunk 1 above)
Comment (jordis-ai2) [reply projects/objectnav_baselines/experiments/robothor/objectnav_robothor_rgb_resnet18gru_ddppo.py]: Now I understand. I think your solution will be cleaner than hacking solutions for each upcoming similar situation 👍
